`Development`: Fix team students eager loading for several occasions

julian-christl opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.
- Removed Atlassian Stack as it's deprecated

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
After issues with #7829 and some hot fixes, I went through the whole code base to find any occurrence where loading team students might be necessary. It took some time, but I was able to go through most branches. There were a few that I had to skip because they were just too complex, but I added an initialisation on these instances, thus they are still covered.

### Description
<!-- Describe your changes in detail -->
Eagerly loaded team students wherever they are called. In some areas I maintained the relations after saving by overriding them afterwards. Some queries I had to extract as `participation.team.students` won't work on the entity graph because team is on the student participation, not the abstract one. In some sections I load the students if not initialised. Usually because it would be too much effort to go through all call branches and make sure that every one of them loads the students for team exercises. Finally I also modified some sections to reduce database calls and improved some queries.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor
- 2 Students in a team
- 1 Exercise of each type as team exercise

1. Ensure that the team administration (create, update, etc) is fully working. You can check this in the test preparation already.
2. Do all steps of the participation (submit, assess, complain, answer complain) and make sure that for every step you check all team views (i.e. the view where team members are listed).
3. Make sure that the scores are displayed correctly at all times for students and instructor alike
4. Make sure that when updating the due date of programming exercises, the repositories get locked/unlocked correctly
5. Make sure that the websocket events work especially for the students and hence the values get updated. If you're not sure what to expect, a good suggestion would be to monitor sentry during your tests and make sure that there are no errors from your server related to this PR (especially initialization errors) https://sentry.ase.in.tum.de/organizations/artemis/issues/?environment=test&project=2&statsPeriod=1h

#### Exam Mode Testing
<!-- If this PR changes some components that are also used in the exam mode, the PR needs additional testing that the exam mode is still working as expected. -->
<!-- If the testing steps above already describe the exam mode or the exam mode cannot be affected by this PR in any way, you can leave this out. -->

Not required, exams don't support team exercises.

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->
No functional changes really affecting the coverage. If you really need the coverage, let me know and I'll add it.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
	- Improved efficiency in fetching related entities for exercises.
	- Enhanced submission query methods for better data handling and added a new method for fetching specific submission associations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

krusche commented:
I am not sure if the PR follows the right approach. It seems we load team and students way too often now. I guess it would be better to only load the team with students, when we actually have a team exercise

krusche commented:
I made sure that the hot fix in the release branch `release/6.7.7` was merged into `develop` to avoid a regression in 6.8.0

julian-christl commented:
> I am not sure if the PR follows the right approach. It seems we load team and students way too often now. I guess it would be better to only load the team with students, when we actually have a team exercise

So do I get you right that you would prefer having two different repository methods for most of these methods and call the corresponding one depending on whether `exercise.isTeamMode()` resolves to false or true?
I don't see the benefit. If it's not a team exercise, it won't get loaded anyway.

Or did I misinterpret your intention here?

julian-christl commented on src/main/java/de/tum/in/www1/artemis/service/scheduled/cache/quiz/QuizScheduleService.java:
```diff
@@ -506,9 +512,10 @@ public void processCachedQuizSubmissions() {
                         else {
-                            if(ltiNewResultService.isPresent()) {
-                                ltiNewResultService.get().onNewResult(participation);
+                            if (participation.getParticipant() instanceof Team team && !Hibernate.isInitialized(team.getStudents())) {
+                                participation.setParticipant(teamRepository.findWithStudentsByIdElseThrow(team.getId()));
                             }
+                            ltiNewResultService.ifPresent(newResultService->newResultService.onNewResult(participation));
```
It's just one additional call in that case and it's a rare endpoint used so there is no performance concern on my side

julian-christl commented on src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:
```diff
@@ -502,16 +505,44 @@ public ResponseEntity<List<StudentParticipation>> updateParticipationDueDates(@P
+            }
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation((ProgrammingExerciseStudentParticipation) participation);
+        });
+    }
```
Fixed in 3e35a0941f104c356dca6c423054b9dad78e67ed

julian-christl commented on src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:
```diff
@@ -502,16 +505,44 @@
+            }
+            programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(exercise, (ProgrammingExerciseStudentParticipation) participation);
+        });
+    }
```
Fixed in 3e35a0941f104c356dca6c423054b9dad78e67ed

undernagruzez reviewed (approved):
code lgtm
tested on ts3
exercise & team creation works as expected
could change the due date to the past and repos got locked
changed the due date back to the future, but the repos remained locks, not sure if it is the expected behavior
did the assessment manually, the students were not notified about the results, also not sure it they should be
also, found these exceptions on sentry: https://sentry.ase.in.tum.de/organizations/artemis/issues/37553/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h
https://sentry.ase.in.tum.de/organizations/artemis/issues/40139/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h
https://sentry.ase.in.tum.de/organizations/artemis/issues/26185/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h

julian-christl commented:
> code lgtm tested on ts3 exercise & team creation works as expected could change the due date to the past and repos got locked changed the due date back to the future, but the repos remained locked, not sure if it is the expected behavior did the assessment manually, the students were not notified about the results, also not sure it they should be also, found these exceptions on sentry: https://sentry.ase.in.tum.de/organizations/artemis/issues/37553/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h https://sentry.ase.in.tum.de/organizations/artemis/issues/40139/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h https://sentry.ase.in.tum.de/organizations/artemis/issues/26185/?environment=test&project=2&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=1h

Thank you for testing
1. Unlocking is not supported, I updated the testing instructions accordingly
2. They should not be notified in terms of a notification, but the result should be updated and reflect the new percentage.
3. I fixed the nullpointer exceptions. I'd appreciate if you could recheck.
4. The last exception is unrelated.

undernagruzez reviewed (approved):
tested again on ts3
didn't encounter any errors

coolchock reviewed (dismissed):
tested everything on ts3, works as expected. It is possible to submit, assess, complain and respond to a complaint. Websockets work too, students get notified about a new assessment. Code LGTM.

rstief commented on src/main/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationService.java:
```diff
@@ -117,6 +118,7 @@ private SingleUserNotification createSingleUserNotification(Object notificationS
      *
      * @param exercise which assessmentDueDate is the trigger for the notification process
      */
+    // TODO: Should by a general method and not be in the single user service
```
do this in a follow-up or create an issue for this, so we don't forget it? :)

rstief reviewed (dismissed):
Code looks good 👍

basak-akan reviewed (dismissed):
Tested on TS3, followed testing steps works as expected

mateusmm01 reviewed (commented):
unrelated

mateusmm01 reviewed (approved):
Tested on ts5 and works as expected 🚀

rstief reviewed (approved):
reapprove

basak-akan reviewed (approved):
Re approve

undernagruzez reviewed (approved):
Re approve
